### PR TITLE
Configurable Lucene analyzer

### PR DESCRIPTION
**Pull request by jackluo923**
Currently, Pinot hard-coded the Lucene analyzer (`standardAnalyzer`) to tokenize strings for indexing and search. In various scenarios, it is extremely useful to customize the analyzer. There are at least two other users who have requested this feature in https://github.com/apache/pinot/issues/9154. 

This PR introduces the capability to specify a custom Lucene analyzer used by text index for indexing and search on an individual column basis. Specifically, this PR allows user to specify the class name (fully qualified) of the Lucene analyzer to use in the text index:

```
fieldConfigList: [
   {
        "name": "columnName",
        "indexType": "TEXT",
        "indexTypes": [
          "TEXT"
        ],
        "properties": {
          "luceneAnalyzerClass": "org.apache.lucene.analysis.core.KeywordAnalyzer"
        },
      }
  ]
  ```    

**Default Behavior**
If user did not specify the `luceneAnalyzerClass` property, the behavior is exactly the same as before which is to use the StandardAnalyzer with couple configuration properties.

**User Specified Behavior 1**
When user specifies the `luceneAnalyzerClass` and the class name is `org.apache.lucene.analysis.standard.StandardAnalyzer`, the behavior is exactly the same as before which is to use the StandardAnalyzer with couple configuration properties.

**User Specified Behavior 2**
When user specifies the `luceneAnalyzerClass` and the class name is not  `org.apache.lucene.analysis.standard.StandardAnalyzer`, the default constructor of the specified Lucene analyzer class is invoked via reflection to create a Lucene analyzer. If user-specified analyzer class does not exist, the ReflectionOperationException is caught and a runtime exception with a more meaningful exception message is thrown.

**Testing**
This configurable Lucene analyzer feature is currently used in production to index and search large amount of text data on multi-variable text columns using `KeywordAnalyzer`. All existing unit tests with the default behavior are passing.

tags: `release-notes`, `enhancement`

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/TextIndexConfig.java [hunk 1]
```diff
@@ -172,6 +187,11 @@ public AbstractBuilder withLuceneMaxBufferSizeMB(int maxBufferSizeMB) {
       _luceneMaxBufferSizeMB = maxBufferSizeMB;
       return this;
     }
+
+    public AbstractBuilder withLuceneAnalyzerFQCN(String luceneAnalyzerFQCN) {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/TextIndexConfig.java:191]:
```suggestion
    public AbstractBuilder withLuceneAnalyzerClass(String luceneAnalyzerClass) {
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/TextIndexUtils.java [hunk 2]
```diff
@@ -112,6 +113,12 @@ private static List<String> parseEntryAsString(@Nullable Map<String, String> col
         .map(String::trim).collect(Collectors.toList());
   }
 
+  public static Analyzer getAnalyzerFromFQCN(@Nonnull String luceneAnalyzerFQCN) throws
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/TextIndexUtils.java:116]:
```suggestion
  public static Analyzer getAnalyzerFromClassName(String luceneAnalyzerClass) throws
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Please update the description to reflect the config key change.

Could you please help add this to the pinot doc after it is merged?

**Comment by jackluo923** [comment]
> LGTM. Please update the description to reflect the config key change.
> 
> Could you please help add this to the pinot doc after it is merged?

Description is updated and I will add the necessary modifications to the Pinot doc repo after this pull request is merged.

**Pull request merged**